neqo-qpack/src/header_block.rs

```diff
@@ -163,25 +121,41 @@ impl HeaderEncoder {
         self.buf.encode_literal(self.use_huffman, NO_PREFIX, value);
     }
 
-    pub fn fix_header_block_prefix(&mut self) {
-        if let Some(r) = self.max_dynamic_index_ref {
+    pub fn encode_header_block_prefix(&mut self) {
+        let tmp = mem::take(&mut self.buf);
+        let (enc_insert_cnt, delta, prefix) = if let Some(r) = self.max_dynamic_index_ref {
             let req_insert_cnt = r + 1;
             if req_insert_cnt <= self.base {
-                self.encode_header_block_prefix(
-                    true,
-                    req_insert_cnt,
+                (
+                    req_insert_cnt % (2 * self.max_entries) + 1,
                     self.base - req_insert_cnt,
-                    true,
-                );
+                    BASE_PREFIX_POSITIVE,
+                )
             } else {
-                self.encode_header_block_prefix(
-                    true,
-                    req_insert_cnt,
+                (
+                    req_insert_cnt % (2 * self.max_entries) + 1,
                     req_insert_cnt - self.base - 1,
-                    false,
-                );
+                    BASE_PREFIX_NEGATIVE,
+                )
             }
-        }
+        } else {
+            (0, self.base, BASE_PREFIX_POSITIVE)
+        };
+        qtrace!(
+            [self],
+            "encode header block prefix max_dynamic_index_ref={:?}, base={}, enc_insert_cnt={}, delta={}, prefix={:?}.",
+            self.max_dynamic_index_ref,
+            self.base,
+            enc_insert_cnt,
+            delta,
+            prefix
+        );
+
+        self.buf
+            .encode_prefixed_encoded_int(NO_PREFIX, enc_insert_cnt);
+        self.buf.encode_prefixed_encoded_int(prefix, delta);
+
+        self.buf.write_bytes(&tmp);
```

Comment:
Not that this needs fixing, but this is worth identifying as a potential performance improvement.

For instance, we could reserve space for the header and write into that space at the end, then return the appropriate slice.